src/ZODB/mvccadapter.py

```diff
@@ -0,0 +1,268 @@
+        'getName', 'getSize', 'history', 'lastTransaction', 'sortKey',
+        'loadBlob', 'openCommittedBlobFile',
+        )
+
+    def __init__(self, storage):
+        self._storage = storage
+        for name in self._copy_methods:
+            if hasattr(self._storage, name):
+                setattr(self, name, getattr(storage, name))
+
+        if interfaces.IBlobStorage.providedBy(storage):
+            zope.interface.alsoProvides(self, interfaces.IBlobStorage)
+
+    def isReadOnly(self):
+        return self._storage.isReadOnly()
+
+    def supportsUndo(self):
+        try:
+            return self._storage.supportsUndo()
+        except AttributeError:
+            return False
+
+    def undoLog(self, first, last, filter=None):
+        try:
+            return self._storage.undoLog(first, last, filter)
+        except AttributeError:
+            return False
+
+    def undoInfo(self, first=0, last=20, specification=None):
+        try:
+            return self._storage.undoInfo(first, last, specification)
+        except AttributeError:
+            return False
+
+    def temporaryDirectory(self):
+        return self._storage.temporaryDirectory()
+
+    def __len__():
+        return len(self._storage)
+
+class MVCCAdapter(Base):
+
+    def __init__(self, storage):
+        Base.__init__(self, storage)
+        self._instances = set()
+        self._lock = threading.Lock()
+        if hasattr(storage, 'registerDB'):
+            storage.registerDB(self)
+
+    def new_instance(self):
+        instance = MVCCAdapterInstance(self)
+        with self._lock:
+            self._instances.add(instance)
+        return instance
+
+    def before_instance(self, before=None):
+        return BeforeAdapterInstance(self, before)
+
+    def undo_instance(self):
+        return UndoAdapterInstance(self)
+
+    def _release(self, instance):
+        with self._lock:
+            self._instances.remove(instance)
+
+    closed = False
+    def close(self):
+        if not self.closed:
+            self.closed = True
+            self._storage.close()
+            del self._instances
+            del self._storage
+
+    def invalidateCache(self):
+        with self._lock:
+            for instance in self._instances:
+                instance._invalidateCache()
+
+    def invalidate(self, transaction_id, oids, version=''):
+        with self._lock:
+            for instance in self._instances:
+                instance._invalidate(oids)
+
+    def _invalidate_finish(self, oids, committing_instance):
+        with self._lock:
+            for instance in self._instances:
+                if instance is not committing_instance:
+                    instance._invalidate(oids)
+
+    references = serialize.referencesf
+    transform_record_data = untransform_record_data = lambda self, data: data
+
+    def pack(self, pack_time, referencesf):
+        return self._storage.pack(pack_time, referencesf)
+
+class MVCCAdapterInstance(Base):
```

Comment:
Well, the only way that code in the wild can look at it is by accessing the connection's underware.  When you do that, you have to expect surprises. :)

These interfaces are a bit of a mess, owing in part to Shane's decision have new_instance clone the existing thing.  IMO, this makes RelStorage's implementation harder to reason about.   <shrug>  Another source of complexity is that both DB exposes a "storage" and connection uses a storage, but what they expose and use are different.  Historically, this didn't matter, because there were the same thing.

If we were really going to work this out, I might have IMVCCStorage and IMVCCConnection with distinct APIs that didn't extend IStorage. I'm willing to do this now, but I'd rather not because I have higher priorities.